### PR TITLE
Add Circuit Generation Notebook for Discrete Time Crystal Experiments

### PR DESCRIPTION
**Pull request by augustehirth**
PR 3 of 6 on implementing and graphing Figures 2d through 3d of Observation of Time-Crystalline Eigenstate Order on a Quantum Processor (arxiv:2107.13571)

This PR adds the notebook which explains the DTC circuit's construction. It depends on PR #250 for the implementation of the circuit constructing function

### CODE REVIEW

**Comment by augustehirth** [comment]
@dstrain115 / @mpharrigan PTAL for method.

@XiaoMiQC  PTAL for correctness.

File: docs/time_crystals/time_crystal_circuit_generation.ipynb [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "7763e5d20492"
+   },
+   "source": [
+    "# Time Crystal Circuit Generation\n",
+    "This notebook covers how Many Body Local Discrete Time Crystal circuit lists are created, from the paper: Observation of Time-Crystalline Eigenstate Order on a Quantum Processor ([Nature](https://www.nature.com/articles/s41586-021-04257-w)). \n",
+    "\n",
+    "Quantum computers and gate-based quantum circuits turn out to be well suited for crafting systems that exhibit time-crystalline behavior. Behavior is crystalline with respect to time if it has some consistent and stable pattern over time. This system's pattern must be resilient against perturbation in the same way that a space-crystalline object, like a diamond, maintains it's structure (is still a diamond) if moved or heated. \n",
```
Comment (dstrain115) [inline docs/time_crystals/time_crystal_circuit_generation.ipynb:35]: "it's structure" should be "its structure"

File: docs/time_crystals/time_crystal_circuit_generation.ipynb [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    "\n",
+    "The quantum computer supplies a system of many qubits, locally connected to each other in a chain. A many-body local system like this is critical for the existence of a time crystal. Without an MBL system, it is expected that the system's state would decay into a maximum entropy state that is incompatible with the goal of stable and consistent time structure. \n",
```
Comment (dstrain115) [inline docs/time_crystals/time_crystal_circuit_generation.ipynb:37]: I would probably expand MBL to many body local or put "A many-body local (MBL)" where it is first mentioned.  Same thing with "DTC"

File: docs/time_crystals/time_crystal_circuit_generation.ipynb [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+   "execution_count": null,
+   "metadata": {
+    "id": "c0f1c1475908"
+   },
+   "outputs": [],
+   "source": [
+    "import inspect\n",
+    "from IPython.display import Code\n",
+    "\n",
+    "Code(inspect.getsource(time_crystals.symbolic_dtc_circuit_list), language=\"python\")"
```
Comment (dstrain115) [inline docs/time_crystals/time_crystal_circuit_generation.ipynb:101]: This seems weird to me that we are printing out the code as an output of the colab.  I'll leave the decision up to you if you stil think this way is clearer, but I would consider linking to the source code with text such as "The code, which can be viewed here (link) is surprisingly succinct."

**Review by dstrain115 — APPROVED** [review]
This seems fine to me modulo some comments.

File: docs/time_crystals/time_crystal_circuit_generation.ipynb (hunk 1 above)
Comment (augustehirth) [reply docs/time_crystals/time_crystal_circuit_generation.ipynb:35]: Ty for catching this, fixed.

File: docs/time_crystals/time_crystal_circuit_generation.ipynb (hunk 2 above)
Comment (augustehirth) [reply docs/time_crystals/time_crystal_circuit_generation.ipynb:37]: Good point, thanks. Fixed.

File: docs/time_crystals/time_crystal_circuit_generation.ipynb (hunk 3 above)
Comment (augustehirth) [reply docs/time_crystals/time_crystal_circuit_generation.ipynb:101]: My intent with printing the code was to provide a detailed example of how the circuits are constructed. Perhaps a couple step-by-step examples of each moment for a circuit of a few qubits would be a better example. I'll try this and see how it looks.

Comment (mpharrigan) [reply docs/time_crystals/time_crystal_circuit_generation.ipynb:101]: I think it's kinda neat 🤷

**Review by mpharrigan — APPROVED** [review]
no complaints from me

**Pull request merged**